src/main/java/build/buildfarm/worker/shard/Worker.java

```diff
@@ -224,6 +227,61 @@ public Worker(String session, ShardWorkerConfig config) throws ConfigurationExce
     this(session, ServerBuilder.forPort(config.getPort()), config);
   }
 
+  /**
+   * The method will prepare the worker for graceful shutdown and send out grpc request to disable
+   * scale in protection when the worker is ready. If unexpected errors happened, it will cancel the
+   * graceful shutdown progress make the worker available again.
+   */
+  public void shutDownWorkerGracefully() {
+    inGracefulShutdown = true;
+    logger.log(Level.INFO, "The current worker is deregistered and should be shutdown gracefully!");
+    int scanRate = 30; // check every 30 seconds
+    int timeWaited = 0;
+    int timeOut = 60 * 15; // 15 minutes
+
+    try {
+      while (!pipeline.isEmpty() && timeWaited < timeOut) {
+        SECONDS.sleep(scanRate);
+        timeWaited += scanRate;
+      }
+    } catch (InterruptedException e) {
+      logger.log(Level.SEVERE, "The worker gracefully shutdown is interrupted: " + e.getMessage());
+    } finally {
+      // make a grpc call to disable scale protection
+      String clusterId = config.getAdminConfig().getClusterId();
```

Comment:
We should check for clusterId first here, if it doesn't exist no point in trying to finish work.